### PR TITLE
Trace exporter for Datadog

### PR DESCRIPTION
**Pull request by nbutton23**
Provides a trace exporter for Datadog. This was closely modeled after the DataDog provided Go exporter.

https://github.com/DataDog/opencensus-go-exporter-datadog

### CODE REVIEW

File: contrib/opencensus-ext-datadog/opencensus/ext/datadog/traces.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+        return self._service
+
+    @property
+    def global_tags(self):
+        """ Specifies the namespaces to which metric keys are appended
+        """
+        return self._global_tags
+
+
+class DatadogTaceExporter(base_exporter.Exporter):
```
Comment (lzchen) [inline contrib/opencensus-ext-datadog/opencensus/ext/datadog/traces.py]: `DatadogTaceExporter` -> `DatadogTraceExporter`

File: contrib/opencensus-ext-datadog/opencensus/ext/datadog/traces.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import defaultdict
+import codecs
+from datetime import datetime
+import bitarray
+from opencensus.common.transports import sync
+from opencensus.trace import base_exporter
+from opencensus.trace import span_data
+from opencensus.ext.datadog.transport import DDTransport
```
Comment (lzchen) [inline contrib/opencensus-ext-datadog/opencensus/ext/datadog/traces.py]: Please sort your imports.

Comment (nbutton23) [reply contrib/opencensus-ext-datadog/opencensus/ext/datadog/traces.py]: I am use to go where first imports are grouped (stdlib, 3rd party, internal) then sorted is the convention for Python the same, or is it just sorted as a unit?

File: contrib/opencensus-ext-datadog/opencensus/ext/datadog/traces.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+        :param list of opencensus.trace.span_data.SpanData span_datas:
+            SpanData tuples to emit
+        """
+        # Map each span data to it's corresponding trace id
+        trace_span_map = defaultdict(list)
+        for sd in span_datas:
+            trace_span_map[sd.context.trace_id] += [sd]
+
+        dd_spans = []
+        # Write spans to Stackdriver
```
Comment (lzchen) [inline contrib/opencensus-ext-datadog/opencensus/ext/datadog/traces.py]: Stackdriver?

File: contrib/opencensus-ext-datadog/opencensus/ext/datadog/traces.py [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+            code = STATUS_CODES.get(span["status"].get("code"))
+            if code is None:
+                code = {}
+                code["message"] = "ERR_CODE_" + str(span["status"].get("code"))
+                code["status"] = 500
+
+            # opencensus.trace.span.SpanKind
+            dd_span['type'] = to_dd_type(span.get("kind"))
+            dd_span["error"] = 0
+            if int(code.get("status") / 100) == 4 or int(
```
Comment (reyang) [inline contrib/opencensus-ext-datadog/opencensus/ext/datadog/traces.py]:
```suggestion
            if 4 <= code.get("status") // 100 <= 5:
```

File: contrib/opencensus-ext-datadog/opencensus/ext/datadog/traces.py [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+    cut_off = len(id_bitarray)
+    if cut_off > 64:
+        cut_off = 64
+    id_cutoff_bytearray = id_bitarray[:cut_off].tobytes()
+    id_int = int(codecs.encode(id_cutoff_bytearray, 'hex'), 16)
+    return id_int
+
+
+# 2019-09-19T14:05:15.808570Z
+TIME_FMT = "%Y-%m-%dT%H:%M:%S.%fZ"
```
Comment (reyang) [inline contrib/opencensus-ext-datadog/opencensus/ext/datadog/traces.py]: Potentially we can reuse the logic: https://github.com/census-instrumentation/opencensus-python/blob/a6d6834974b4ebf273905e2d903f3f1c39cc8e4e/opencensus/common/utils/__init__.py#L29

File: contrib/opencensus-ext-datadog/opencensus/ext/datadog/traces.py [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+
+# https://opencensus.io/tracing/span/status/
+STATUS_CODES = {
+    0: {
+        "message": "OK",
+        "status": int(200)
```
Comment (reyang) [inline contrib/opencensus-ext-datadog/opencensus/ext/datadog/traces.py]: curious - what do we put `int(200)` instead of `200`?

**Comment by reyang** [comment]
Please clean the CI.

File: contrib/opencensus-ext-datadog/opencensus/ext/datadog/traces.py (hunk 6 above)
Comment (nbutton23) [reply contrib/opencensus-ext-datadog/opencensus/ext/datadog/traces.py]: I think that was from when I was having issues with python  int / int = float (I just learned the // operator.) I will clean that up.

**Comment by nbutton23** [comment]
@reyang I am new to python and might need a little help with understanding how to fix the travis-ci.  Running nox locally works. . .

**Comment by lzchen** [comment]
@nbutton23 
We use `nox` to run our CircleCI build, while `tox` is responsible for our TravisCI build (confusing I know). Taking a look at the errors from your build, it looks like it cannot find the dependency for your extension when running the tests in the virtual environment. Installing the datadog extension should solve it.

Add `py{27,34,35,36,37}-unit,py37-lint: -e contrib/opencensus-ext-datadog` to `tox.ini` file under `deps`.

File: contrib/opencensus-ext-datadog/opencensus/ext/datadog/traces.py (hunk 6 above)
Comment (nbutton23) [reply contrib/opencensus-ext-datadog/opencensus/ext/datadog/traces.py]: cleaned up

**Comment by reyang** [comment]
Thanks for your contribution @nbutton23!
Welcome to Python and OpenCensus 😄

**Comment by reyang** [comment]
@nbutton23 it'll be great to have an `README.rst` file as other extension packages do. This file will be published to PyPI so people can see the docs. You may refer to https://pypi.org/project/opencensus-ext-azure/ and https://github.com/census-instrumentation/opencensus-python/tree/master/contrib/opencensus-ext-azure.

**Comment by nbutton23** [comment]
@reyang I opened a new PR for the docs https://github.com/census-instrumentation/opencensus-python/pull/799

**Pull request merged**